src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -155,6 +162,60 @@ public async Task<WorkItem> CreateRcaWorkItem(string project, string title, Canc
             return JsonConvert.DeserializeObject<WorkItem>(json);
         }
 
+        public delegate bool TryParseLine(string line, out string parsedValue);
+
+        public async Task<HttpResponseMessage> TryGetLogContents(
+            string logUri,
+            CancellationToken cancellationToken)
+        {
+            for (int i = 0; i < _retryNumber; i++)
+            {
+                using (var request = new HttpRequestMessage(HttpMethod.Get, logUri))
+                {
+                    request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/plain"));
+
+                    try
+                    {
+                        var response = await _httpClient.SendAsync(request, cancellationToken);
+
+                        if (response.StatusCode == HttpStatusCode.TooManyRequests)
+                        {
+                            if (response.Headers.TryGetValues("Retry-After", out var values) &&
+                                int.TryParse(values.First(), out var seconds))
+                            {
+                                int randomDelay;
+                                lock (_randomNumberGenerator)
+                                {
+                                    randomDelay = _randomNumberGenerator.Next(_retryDelayMin, _retryDelayMax);
+                                }
+                                await Task.Delay(TimeSpan.FromSeconds(seconds).Add(TimeSpan.FromMilliseconds(randomDelay)), cancellationToken);
+                                continue;
+                            }
+                        }
+
+                        response.EnsureSuccessStatusCode();
+                        return response;
+                    }
+                    catch (OperationCanceledException ex) when (ex.CancellationToken == cancellationToken)
+                    {
+                        //Exit if the task got cancelled
+                        throw;
+                    }
+                    catch (Exception)
+                    {
+                        int randomDelay;
+                        lock (_randomNumberGenerator)
+                        {
+                            randomDelay = _randomNumberGenerator.Next(_retryDelayMin, _retryDelayMax);
+                        }
+                        await Task.Delay(TimeSpan.FromMilliseconds(randomDelay));
+                    }
+                }
+            }
+
+            throw new InvalidOperationException($"Failed to get logs after retrying {_retryNumber} times");
```

Comment:
Not sure `InvalidOperationException` is the right one. Maybe just `Exception`?

```suggestion
            throw new Exception($"Failed to get logs after retrying {_retryNumber} times");
```

Reply:
@lpatalas was telling me that it's not the best idea to throw the base exception

Reply:
Ok, then some HTTP or web request exception?

Reply:
The code should jsut throw whatever exception caused it to fail, rather than swallowing that exception and trying to create a new one.  Or use one of the existing retry classes that handles all this for us.